cloud/scope/vpc_cluster.go

```diff
@@ -486,9 +497,186 @@ func (s *VPCClusterScope) createVPC() (*vpcv1.VPC, error) {
+		ID:   *image.ID,
+		Name: image.Name,
+		// We must wait for the image to be ready, on followup reconciliation loops.
+		Ready: false,
+	})
+	return true, nil
+}
+
+// createCustomImage will create a new VPC Custom Image.
+func (s *VPCClusterScope) createCustomImage() (*vpcv1.Image, error) {
+	if s.IBMVPCCluster.Spec.Image == nil {
+		return nil, fmt.Errorf("error failed to create vpc custom image, no image spec defined")
+	}
+
+	// Collect the Resource Group ID.
+	var resourceGroupID *string
+	// Check Resource Group in Image spec.
+	if s.IBMVPCCluster.Spec.Image.ResourceGroup != nil {
+		if s.IBMVPCCluster.Spec.Image.ResourceGroup.ID != "" {
+			resourceGroupID = ptr.To(s.IBMVPCCluster.Spec.Image.ResourceGroup.ID)
+		} else if s.IBMVPCCluster.Spec.Image.ResourceGroup.Name != nil {
+			id, err := s.ResourceManagerClient.GetResourceGroupByName(*s.IBMVPCCluster.Spec.Image.ResourceGroup.Name)
+			if err != nil {
+				return nil, fmt.Errorf("error retrieving resource group by name: %w", err)
+			}
+			resourceGroupID = id.ID
+		}
+	} else {
+		// Otherwise, we will use the cluster Resource Group ID, as we expect to create all resources in that Resource Group.
+		id, err := s.GetResourceGroupID()
+		if err != nil {
+			return nil, fmt.Errorf("error retrieving resource group id: %w", err)
+		}
+		resourceGroupID = ptr.To(id)
+	}
+
+	// We must have an OperatingSystem value supplied in order to create the Custom Image.
+	// NOTE(cjschaef): Perhaps we could try defaulting this value, so it isn't required for Custom Image creation.
+	if s.IBMVPCCluster.Spec.Image.OperatingSystem == nil {
+		return nil, fmt.Errorf("error failed to create vpc custom image due to missing operatingSystem")
+	}
+
+	// Build the COS Object URL using the ImageSpec
+	fileHRef, err := s.buildCOSObjectHRef()
+	if err != nil {
+		return nil, fmt.Errorf("error building vpc custom image file href: %w", err)
+	} else if fileHRef == nil {
+		return nil, fmt.Errorf("error failed to build vpc custom image file href")
+	}
+
+	options := &vpcv1.CreateImageOptions{
+		ImagePrototype: &vpcv1.ImagePrototype{
+			Name: s.IBMVPCCluster.Spec.Image.Name,
+			File: &vpcv1.ImageFilePrototype{
+				Href: fileHRef,
+			},
+			OperatingSystem: &vpcv1.OperatingSystemIdentity{
+				Name: s.IBMVPCCluster.Spec.Image.OperatingSystem,
+			},
+			ResourceGroup: &vpcv1.ResourceGroupIdentity{
+				ID: resourceGroupID,
+			},
+		},
+	}
+
+	imageDetails, _, err := s.VPCClient.CreateImage(options)
+	if err != nil {
+		return nil, fmt.Errorf("error unknown failure creating vpc custom image: %w", err)
+	}
+	if imageDetails == nil || imageDetails.ID == nil || imageDetails.Name == nil || imageDetails.CRN == nil {
+		return nil, fmt.Errorf("error failed creating custom image")
+	}
+
+	// NOTE: This tagging is only attempted once. We may wish to refactor in case this single attempt fails.
+	if err := s.TagResource(s.Name(), *imageDetails.CRN); err != nil {
```

Comment:
If tagging fails, after creating the Custom Image (or any resource with this similar logic, like VPC), we return an error. Next loop through could then find the resource that was created the first attempt, and update the resource Status (in the VPC example, a VPC is likely "ready" in the second attempt), with the expectation it is found by the name used during creation. This will monitor and/or update the resource Status accordingly. As we expect to not try to recreate a resource that was already created, we don't expect to try to retag the resource.

Logic could be added to check tags and apply if missing, but that logic will be more cumbersome and time consuming that desired currently, and I only plan to worry about tags during Delete actions, which are a secondary concern currently.

Reply:
I can see you have handled this in ReconcileVPC, but here, you don't have a func like GetVPCID. You are checking either from status or from spec crn. If it's there you are validating the image and set the status based on that. Please take care of this here.

Reply:
Not sure I am following, but if the recommendation here is to update status (SetResourceStatus) prior to tagging, I can move that for Custom Images.

I can handle VPC in a similar fashion then, in a separate PR I expect.